score.go

```diff
@@ -0,0 +1,885 @@
+	if !ok {
+		return
+	}
+
+	tstats, ok := pstats.getTopicStats(topic, ps.params)
+	if !ok {
+		return
+	}
+
+	tstats.inMesh = true
+	tstats.graftTime = time.Now()
+	tstats.meshTime = 0
+	tstats.meshMessageDeliveriesActive = false
+}
+
+func (ps *peerScore) Prune(p peer.ID, topic string) {
+	ps.Lock()
+	defer ps.Unlock()
+
+	pstats, ok := ps.peerStats[p]
+	if !ok {
+		return
+	}
+
+	tstats, ok := pstats.getTopicStats(topic, ps.params)
+	if !ok {
+		return
+	}
+
+	// sticky mesh delivery rate failure penalty
+	threshold := ps.params.Topics[topic].MeshMessageDeliveriesThreshold
+	if tstats.meshMessageDeliveriesActive && tstats.meshMessageDeliveries < threshold {
+		deficit := threshold - tstats.meshMessageDeliveries
+		tstats.meshFailurePenalty += deficit * deficit
+	}
+
+	tstats.inMesh = false
+}
+
+func (ps *peerScore) ValidateMessage(msg *Message) {
+	ps.Lock()
+	defer ps.Unlock()
+
+	// the pubsub subsystem is beginning validation; create a record to track time in
+	// the validation pipeline with an accurate firstSeen time.
+	_ = ps.deliveries.getRecord(ps.msgID(msg.Message))
+}
+
+func (ps *peerScore) DeliverMessage(msg *Message) {
+	ps.Lock()
+	defer ps.Unlock()
+
+	ps.markFirstMessageDelivery(msg.ReceivedFrom, msg)
+
+	drec := ps.deliveries.getRecord(ps.msgID(msg.Message))
+
+	// mark the message as valid and reward mesh peers that have already forwarded it to us
+	drec.status = delivery_valid
+	drec.validated = time.Now()
+	for p := range drec.peers {
+		// this check is to make sure a peer can't send us a message twice and get a double count
+		// if it is a first delivery.
+		if p != msg.ReceivedFrom {
+			ps.markDuplicateMessageDelivery(p, msg, time.Time{})
+		}
+	}
+}
+
+func (ps *peerScore) RejectMessage(msg *Message, reason string) {
+	ps.Lock()
+	defer ps.Unlock()
+
+	// TODO: the reasons should become named strings; good enough for now.
+	switch reason {
+	// we don't track those messages, but we penalize the peer as they are clearly invalid
+	case "missing signature":
+		fallthrough
+	case "invalid signature":
+		ps.markInvalidMessageDelivery(msg.ReceivedFrom, msg)
+		return
+
+		// we ignore those messages, so do nothing.
+	case "blacklisted peer":
+		fallthrough
+	case "blacklisted source":
+		return
+	}
+
+	drec := ps.deliveries.getRecord(ps.msgID(msg.Message))
+
+	if reason == "validation throttled" {
+		// if we reject with "validation throttled" we don't penalize the peer(s) that forward it
+		// because we don't know if it was valid.
+		drec.status = delivery_throttled
+		// release the delivery time tracking map to free some memory early
+		drec.peers = nil
+		return
+	}
+
+	// mark the message as invalid and penalize peers that have already forwarded it.
+	drec.status = delivery_invalid
+	for p := range drec.peers {
+		ps.markInvalidMessageDelivery(p, msg)
+	}
+
+	// release the delivery time tracking map to free some memory early
+	drec.peers = nil
+}
+
+func (ps *peerScore) DuplicateMessage(msg *Message) {
```

Comment:
@vyzo I'm getting some "assignment to nil map" errors running against this branch in testground, e.g.

```
   2.0120s      ERROR << c694df9af4d2 >> panic: assignment to entry in nil map
   2.0121s      ERROR << c694df9af4d2 >>
   2.0121s      ERROR << c694df9af4d2 >> goroutine 98 [running]:
   2.0121s      ERROR << c694df9af4d2 >> github.com/libp2p/go-libp2p-pubsub.(*peerScore).DuplicateMessage(0xc000091340, 0xc00025ebd0)
   2.0122s      ERROR << c694df9af4d2 >> 	/go/pkg/mod/github.com/libp2p/go-libp2p-pubsub@v0.2.7-0.20200311004457-925ed3894d0b/score.go:653 +0x14f
   2.0122s      ERROR << c694df9af4d2 >> github.com/libp2p/go-libp2p-pubsub.(*pubsubTracer).DuplicateMessage(0xc000091300, 0xc00025ebd0)
   2.0122s      ERROR << c694df9af4d2 >> 	/go/pkg/mod/github.com/libp2p/go-libp2p-pubsub@v0.2.7-0.20200311004457-925ed3894d0b/trace.go:106 +0x3a1
   2.0122s      ERROR << c694df9af4d2 >> github.com/libp2p/go-libp2p-pubsub.(*validation).validate(0xc000389290, 0x0, 0x0, 0x0, 0xc0000a8f90, 0x26, 0xc00025ebd0)
   2.0123s      ERROR << c694df9af4d2 >> 	/go/pkg/mod/github.com/libp2p/go-libp2p-pubsub@v0.2.7-0.20200311004457-925ed3894d0b/validation.go:207 +0x6ad
   2.0133s      ERROR << c694df9af4d2 >> github.com/libp2p/go-libp2p-pubsub.(*validation).validateWorker(0xc000389290)
   2.0133s      ERROR << c694df9af4d2 >> 	/go/pkg/mod/github.com/libp2p/go-libp2p-pubsub@v0.2.7-0.20200311004457-925ed3894d0b/validation.go:184 +0x7c
   2.0133s      ERROR << c694df9af4d2 >> created by github.com/libp2p/go-libp2p-pubsub.(*validation).Start
   2.0134s      ERROR << c694df9af4d2 >> 	/go/pkg/mod/github.com/libp2p/go-libp2p-pubsub@v0.2.7-0.20200311004457-925ed3894d0b/validation.go:98 +0x70
   2.3247s INCOMPLETE << c694df9af4d2 >>
```

This must be because `drec.peers` was released in an earlier `RejectMessage` call. Can you add a nil check here for `drec.peers` and exit early if it's already been rejected? Or do we actually want to keep tracking duplicate deliveries after we've rejected the message?

Reply:
hrm, we do want to keep track of rejected duplicates, but I am not sure this is the case here.
Are you generating any rejected messages?

Reply:
It's in a valid delivery.

Reply:
I think we might have some throttling going on which can also set the peers map to nil.

Reply:
ie, it is possible to start validation and also throttle validation of a duplicate because it can't push to the validation workers.

Reply:
yeah, that's what I think is happening: the validation queue throttles before we enter validation, which rejects the message (with a message ID of unknown validity) and sets the peer map to nil.
In the meantime there is another validation happening that completes successfully, and boom.
Anyhow, a nasty little bug.

Reply:
it should be fixed in https://github.com/libp2p/go-libp2p-pubsub/pull/263/commits/5f1b070d5922dc076572cbd03010766a6f8b89cf